INT-3852: Support ObjectName Customization

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3852

Custom object names in `@IntegrationManagedResource` were ignored.

Some EIP objects were wrapped with their `Lifecycle` in order to expose
a single MBean with all attributes and operations.

This process "hid" the annotation so the object namer failed to detect the
presence of a custom object name or the other attributes.

Also, update the `@IntegrationManagedResource` to use the Spring 4.2 `@AliasFor` annotation.

garyrussell committed:
INT-3852: Support ObjectName Customization

JIRA: https://jira.spring.io/browse/INT-3852

Custom object names in `@IntegrationManagedResource` were ignored.

Some EIP objects were wrapped with their `Lifecycle` in order to expose
a single MBean with all attributes and operations.

This process "hid" the annotation so the object namer failed to detect the
presence of a custom object name or the other attributes.

Also, update the `@IntegrationManagedResource` to use the Spring 4.2 `@AliasFor` annotation.

artembilan commented on spring-integration-jmx/src/main/java/org/springframework/integration/monitor/IntegrationMBeanExporter.java:
```diff
@@ -1067,6 +1073,21 @@ private static Object getField(Object target, String name) {
 
+	private org.springframework.jmx.export.metadata.ManagedResource getManagedResourceForTarget(
+			Object managedBean) {
+		DirectFieldAccessor accessor = new DirectFieldAccessor(managedBean);
```
Don't see reason to create `DFA`, if the next `if` doesn't pass.

artembilan committed:
INT-3852: Polishing

artembilan commented:
@garyrussell , please, take a look to my polishing on the matter: https://github.com/artembilan/spring-integration/commit/baa5789619ce98555b0f5b1382dff5939f773f57.

I tried to reuse the "super" code as much as possible.

garyrussell committed:
Polishing - Test with Proxy

garyrussell commented:
Nice polishing; looks good; much cleaner; thanks!!

I was concerned about the delegate being a JDK proxy (`getClassToExpose()` only unwraps CGLib) ; but I see we already extracted the target before setting up the delegate in the wrapper.

I went ahead and added a test for a proxied handler, though.

artembilan commented:
Agreed. That exactly would be better to have that test as well :smile:

artembilan commented:
Cool!
Merging...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/2fb039004d844159bc389710c36fe8798dafc09d
